kml.cc

```diff
@@ -448,15 +458,51 @@ void gx_trk_s(xg_string, const QXmlStreamAttributes*)
   track_add_head(gx_trk_head);
   delete gx_trk_times;
   gx_trk_times = new QList<gpsbabel::DateTime>;
+  delete gx_trk_coords;
+  gx_trk_coords = new QList<std::tuple<int, double, double, double>>;
 }
 
 void gx_trk_e(xg_string, const QXmlStreamAttributes*)
 {
+  // Check that for every temporal value (kml:when) in a kml:Track there is a position (kml:coord) value.
+  // Check that for every temporal value (kml:when) in a gx:Track there is a position (gx:coord) value.
+  if (gx_trk_times->size() != gx_trk_coords->size()) {
+    fatal(MYNAME ": There were more coord elements than the number of when elements.\n");
+  }
+
+  // In KML 2.3 kml:Track elements kml:coord and kml:when elements are not required to be in any order.
+  // In gx:Track elements all kml:when elements are required to precede all gx:coord elements.
+  // For both we allow any order.  Many writers using gx:Track elements don't adhere to the schema.
+  while (!gx_trk_times->isEmpty()) {
+    Waypoint* trkpt = new Waypoint;
+    trkpt->SetCreationTime(gx_trk_times->takeFirst());
+    double lat, lon, alt;
+    int n;
+    std::tie(n, lat, lon, alt) = gx_trk_coords->takeFirst();
+    // An empty kml:coord element is permitted to indicate missing position data;
+    // the estimated position may be determined using some interpolation method.
+    // However if we get one we will throw away the time as we don't have a location.
+    // It is not clear that coord elements without altitude are allowed, but our
```

Comment:
That's the KML equivalent of (and at some level, the inspiration for) our "unknown altitude" values. In Earth, it mostly means 'clamped to ground' or 'draped over terrain' to stop things like boat rides dipping below water when the number is likely to be close to zero vs. being legitimately unknown. We (GPSBabel) don't handle this consistently.